// @hasezoey/typegoose/lib/internal/errors.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const utils_1 = require("./utils");
class InvalidPropError extends Error {
    constructor(targetName, key) {
        super(`Property "${targetName}.${key}" has no buildSchema!`);
    }
}
exports.InvalidPropError = InvalidPropError;
class InvalidTypeError extends Error {
    constructor(targetName, key, Type) {
        super(`"${targetName}.${key}"'s Type is invalid! Type is: "${Type}"`);
    }
}
exports.InvalidTypeError = InvalidTypeError;
class NotNumberTypeError extends Error {
    constructor(key) {
        super(`Type of ${key} property is not a number.`);
    }
}
exports.NotNumberTypeError = NotNumberTypeError;
class NotStringTypeError extends Error {
    constructor(key) {
        super(`Type of ${key} property is not a string.`);
    }
}
exports.NotStringTypeError = NotStringTypeError;
class NoMetadataError extends Error {
    constructor(key) {
        super(`There is no metadata for the "${key}" property.\n` +
            'Check if emitDecoratorMetadata is enabled in tsconfig.json ' +
            'or check if you\'ve declared a sub document\'s class after usage.');
    }
}
exports.NoMetadataError = NoMetadataError;
/** Not All Virtual Populate Elements Error */
class NotAllVPOPElementsError extends Error {
    constructor(name, key) {
        super(`"${name}.${key}" has not all needed Virtual Populate Options! Needed are: ${utils_1.allVirtualoptions.join(', ')}`);
    }
}
exports.NotAllVPOPElementsError = NotAllVPOPElementsError;
class NoValidClass extends TypeError {
    constructor(cl) {
        super(`"${cl}" is not a function(/constructor)!`);
    }
}
exports.NoValidClass = NoValidClass;
