// @hasezoey/typegoose/lib/internal/utils.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const util_1 = require("util");
const data_1 = require("./data");
const primitives = ['String', 'Number', 'Boolean', 'Date', 'Decimal128', 'ObjectID', 'Array'];
/**
 * Returns true, if it includes the Type
 * @param Type The Type
 * @returns true, if it includes it
 */
function isPrimitive(Type) {
    return primitives.includes(Type.name);
}
exports.isPrimitive = isPrimitive;
/**
 * Returns true, if it is an Object
 * @param Type The Type
 * @returns true, if it is an Object
 */
function isObject(Type) {
    let prototype = Type.prototype;
    let name = Type.name;
    while (name) {
        if (name === 'Object') {
            return true;
        }
        prototype = Object.getPrototypeOf(prototype);
        name = prototype ? prototype.constructor.name : null;
    }
    return false; // can this even return false?
    // return !isNullOrUndefined(Type) && (typeof Type === 'object' || Type.name === 'Object') && !Array.isArray(Type);
}
exports.isObject = isObject;
/**
 * Returns true, if it is an Number
 * @param Type The Type
 * @returns true, if it is an Number
 */
function isNumber(Type) {
    return Type.name === 'Number';
}
exports.isNumber = isNumber;
/**
 * Returns true, if it is an String
 * @param Type The Type
 * @returns true, if it is an String
 */
function isString(Type) {
    return Type.name === 'String';
}
exports.isString = isString;
/**
 * Initialize as Object
 * @param name The Name of the Schema
 * @param key The Property key to set
 */
function initAsObject(name, key) {
    if (!data_1.schemas.get(name)) {
        data_1.schemas.set(name, {});
    }
    data_1.schemas.get(name)[key] = {};
}
exports.initAsObject = initAsObject;
/**
 * Initialize as Array
 * @param name The Name of the Schema
 * @param key The Property key to set
 */
function initAsArray(name, key) {
    if (!data_1.schemas.get(name)) {
        data_1.schemas.set(name, {});
    }
    data_1.schemas.get(name)[key] = [{}];
}
exports.initAsArray = initAsArray;
/**
 * Get the Class for a given Document
 * @param document The Document
 */
function getClassForDocument(document) {
    const modelName = document.constructor.modelName;
    return data_1.constructors.get(modelName);
}
exports.getClassForDocument = getClassForDocument;
/**
 * Return true if there are Options
 * @param options The raw Options
 */
function isWithStringValidate(options) {
    return !util_1.isNullOrUndefined(options.match
        || options.enum
        || options.minlength
        || options.maxlength);
}
exports.isWithStringValidate = isWithStringValidate;
/**
 * Return true if there are Options
 * @param options The raw Options
 */
function isWithStringTransform(options) {
    return !util_1.isNullOrUndefined(options.lowercase || options.uppercase || options.trim);
}
exports.isWithStringTransform = isWithStringTransform;
/**
 * Return true if there are Options
 * @param options The raw Options
 */
function isWithNumberValidate(options) {
    return !util_1.isNullOrUndefined(options.min || options.max);
}
exports.isWithNumberValidate = isWithNumberValidate;
const virtualOptions = ['localField', 'foreignField'];
/**
 * Check if Options include Virtual Populate Options
 * @param options RawOptions of the Prop
 */
function isWithVirtualPOP(options) {
    return Object.keys(options).some((v) => virtualOptions.includes(v));
}
exports.isWithVirtualPOP = isWithVirtualPOP;
exports.allVirtualoptions = virtualOptions.slice(0);
exports.allVirtualoptions.push('ref');
/**
 * Check if All the required Options are present
 * @param options RawOptions of the Prop
 */
function includesAllVirtualPOP(options) {
    return exports.allVirtualoptions.every((v) => Object.keys(options).includes(v));
}
exports.includesAllVirtualPOP = includesAllVirtualPOP;
/**
 * Check if the given value has options of "IModelOptions"
 * @param value The Value to evaulate
 * @internal
 */
function isModelOptions(value) {
    return value && (typeof value.schemaOptions === 'object' ||
        typeof value.options === 'object');
}
/**
 * Merges existing metadata with new value
 * @param key Metadata key
 * @param value Raw value
 * @param cl The constructor
 * @internal
 */
function assignMetadata(key, value, cl) {
    const current = Reflect.getMetadata(key, cl) || {};
    // the following checks are needed, so that the new value dosnt override the full options
    // "deepmerge" cannot be used because of the other options like "existingMongoose"
    if (isModelOptions(value) && !util_1.isNullOrUndefined(current.schemaOptions)) {
        value.schemaOptions = Object.assign(current.schemaOptions, value.schemaOptions);
    }
    if (isModelOptions(value) && !util_1.isNullOrUndefined(current.options)) {
        value.options = Object.assign(current.options, value.options);
    }
    const newValue = Object.assign(current, value);
    Reflect.defineMetadata(key, newValue, cl);
}
exports.assignMetadata = assignMetadata;
/**
 * Get the correct name of the class's model
 * (with suffix)
 * @param cl The Class
 */
function getName(cl) {
    // disabled until hasezoey#23 & hasezoey#24 gets fixed
    // const options: IModelOptions = Reflect.getMetadata(DecoratorKeys.ModelOptions, cl) || {};
    // const baseName = cl.name;
    // const suffix = (options.options ? options.options.customName : undefined) ||
    //   (options.schemaOptions ? options.schemaOptions.collection : undefined);
    // return suffix ? `${baseName}_${suffix}` : baseName;
    return cl.name;
}
exports.getName = getName;
